src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala

```diff
@@ -39,47 +41,47 @@ case class IllegalLabelKeyException(labelKey: String)
   extends LeoException(s"Labels cannot have a key of '$labelKey'", StatusCodes.NotAcceptable)
 
 
-class LeonardoService(protected val dataprocConfig: DataprocConfig, protected val clusterResourcesConfig: ClusterResourcesConfig, protected val proxyConfig: ProxyConfig, gdDAO: DataprocDAO, dbRef: DbReference, val clusterMonitorSupervisor: ActorRef)(implicit val executionContext: ExecutionContext) extends LazyLogging {
+class LeonardoService(protected val dataprocConfig: DataprocConfig,
+                      protected val clusterResourcesConfig: ClusterResourcesConfig,
+                      protected val proxyConfig: ProxyConfig,
+                      protected val petServiceAccountConfig: PetServiceAccountConfig,
+                      protected val gdDAO: DataprocDAO,
+                      protected val googleIamDAO: GoogleIamDAO,
+                      protected val dbRef: DbReference,
+                      protected val clusterMonitorSupervisor: ActorRef,
+                      protected val samDAO: SamDAO)
+                     (implicit val executionContext: ExecutionContext) extends LazyLogging {
   private val bucketPathMaxLength = 1024
   private val includeDeletedKey = "includeDeleted"
 
   // Register this instance with the cluster monitor supervisor so our cluster monitor can potentially delete and recreate clusters
   clusterMonitorSupervisor ! RegisterLeoService(this)
 
-  def createCluster(googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest): Future[Cluster] = {
-    def create() = {
-      val clusterRequestWithDefaults = processClusterRequest(googleProject, clusterName, clusterRequest)
-      createGoogleCluster(googleProject, clusterName, clusterRequestWithDefaults) flatMap { case (cluster: Cluster, initBucket: GcsBucketName) =>
-        dbRef.inTransaction { dataAccess =>
-          dataAccess.clusterQuery.save(cluster, GcsPath(initBucket, GcsRelativePath("")))
-        }
-      }
+  def createCluster(userInfo: UserInfo, googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest): Future[Cluster] = {
+    samDAO.getPetServiceAccount(userInfo).flatMap { serviceAccount =>
+      createCluster(serviceAccount, googleProject, clusterName, clusterRequest)
     }
+  }
 
+  def createCluster(serviceAccount: WorkbenchUserServiceAccountEmail, googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest): Future[Cluster] = {
```

Comment:
These 2 variants of `createCluster` exist to support cluster recreation-on-failure. On cluster recreation, we already have the service account stored in the DB, so we don't need to call Sam again.

Reply:
Could you add comments for the two variants?